### ISSUE TITLE
Unexpected results for `fetchCalendarObjects` when using rrule expansion

### ISSUE THREAD

**Issue by dpraul** [natelindev/tsdav#116]
Hi there! I'm getting unexpected results when using `fetchCalendarObject`, and I'm not certain if it's something wrong with the library or due to my own lack of knowledge on iCal. If it matters, I am this to interface with a local [DAViCal](https://davical.org/) instance that is otherwise empty.

I have created an empty calendar and added one event that recurs weekly. When I query for a date-range that does _not_ overlap with a recurrence of the item, I am still getting that event as a result - where I would expect to get no results. When I use the `expand` flag, the results are only an empty VCALENDAR. The following sample code demonstrates the behavior:

```js
// make a new (empty!) calendar
const calendar = await client.makeCalendar({
  url: 'url-for-calendar', props: { displayname: 'Test Calendar' },
});

// create an event that recurs weekly
const eventId = 'f2af973a-c9cc-4410-ac79-c7c32fad70e7';
const iCalString = `
BEGIN:VCALENDAR
BEGIN:VEVENT
SUMMARY:Test Event
UID:${eventId}
DTSTART:20220606T000000
RRULE:FREQ=WEEKLY
END:VEVENT
END:VCALENDAR
`.trim();
await client.createCalendarObject({
  calendar, iCalString, filename: `${eventId}.ics`,
});

// query a range OUTSIDE a recurrence of the event
const events = await client.fetchCalendarObjects({
  calendar,
  expand: true,
  timeRange: {
    start: '2022-06-08T10:00:00.000Z',
    end: '2022-06-09T10:00:00.000Z',
  },
});
// RESULTS:
// events = [{
//   url: 'url-for-calendar/f2af973a-c9cc-4410-ac79-c7c32fad70e7.ics',
//   etag: '"f1569d1a6f2562ce9cd486e4b3ef1d09"',
//   data: 'BEGIN:VCALENDAR\r\nEND:VCALENDAR'
// }]
```

**Comment by natelindev** [comment]
Thanks for reporting, I'll investigate it this week

**Comment by natelindev** [comment]
> I did some more digging here and think I'm onto something, but excuse my naïveté - I'm very much shooting in the dark here.
> 
> I suspect my results include the unexpected event because the expand [isn't factored in during the `calendarQuery`](https://github.com/natelindev/tsdav/blob/f366a05d5f27ad742401ae93cfd6de93c7645002/src/calendar.ts#L247) before the `calendarMultiGet`. I'm not certain _why_ DAViCal returns the event even though it doesn't match the range - probably some nuance of how it handles non-expanded recurring events - but it does _not_ get returned if the expand is included in the props.
> 
> That said, it appears I can replace the call to `fetchCalendarObjects` with a single `calendarQuery` using getetag and expand in the props + time-range in the filters and I get the results I would expect - correctly matching events for the date range, without empty objects. That said, I do not know why the original command needs to perform two requests, so I may be shooting myself in the foot by doing this.
> 
> I'll do more testing here, but if that works then my problem is solved - I would just use `calendarQuery` instead of the `fetchCalendarObjects` utility.

Sorry for waiting so long before my update, the reason for using `calendarQuery` and `calendarMultiGet` together is that we need to first query for the calendar object urls to fetch, then actually fetching the data using calendarMulti fetch. The reason for this is some CALDAV prodivders do not support querying calendar data directly using `calendarQuery`.

That being said, the point of exposing all functions used by the lib is that you can always combine them to build your own function. Hopefully using calendarQuery directly did fix your problem.

Not taking expand into consideration for calendarQuery is indeed the bug here, I will post an update later to address this issue.

**Comment by natelindev** [comment]
fixed in v2.0.3

**Comment by dpraul** [comment]
> The reason for this is some CALDAV prodivders do not support querying calendar data directly using `calendarQuery`.

This is great to know! I'll keep that in mind for if we change providers.

> That being said, the point of exposing all functions used by the lib is that you can always combine them to build your own function. Hopefully using calendarQuery directly did fix your problem.

Yes, we did manage to get this working just using `calendarQuery`. I'm very pleased that the library is flexible enough to accommodate this :)

> Not taking expand into consideration for calendarQuery is indeed the bug here, I will post an update later to address this issue.

Thanks for looking into it! And thanks as well for adding explicit tests for DAViCal support!
